packages/headless/src/index.ts

```diff
@@ -412,8 +371,59 @@ export default async (opts: HeadlessOptions) => {
       storeDir: opts.storeDir,
       virtualStoreDir,
     })
-  }
 
+    await linkAllBins(graph, { optional: opts.include.optionalDependencies, warn })
+
+    /** Skip linking and writing lockfile due to no project manifest, opts.projects may contains no enough info */
+    if (!opts.ignorePackageManifest) {
+      await Promise.all(opts.projects.map(async ({ rootDir, id, manifest, modulesDir }) => {
+        if (opts.symlink !== false) {
+          await linkRootPackages(filteredLockfile, {
+            importerId: id,
+            importerModulesDir: modulesDir,
+            lockfileDir,
+            projectDir: rootDir,
+            projects: opts.projects,
+            registries: opts.registries,
+            rootDependencies: directDependenciesByImporterId[id],
+          })
+        }
+
+        // Even though headless installation will never update the package.json
+        // this needs to be logged because otherwise install summary won't be printed
+        packageManifestLogger.debug({
+          prefix: rootDir,
+          updated: manifest,
+        })
+      }))
+
+      await Promise.all(opts.projects.map(async (project) => {
+        if (opts.publicHoistPattern?.length && path.relative(opts.lockfileDir, project.rootDir) === '') {
+          await linkBinsOfImporter(project)
+        } else {
+          const directPkgDirs = Object.values(directDependenciesByImporterId[project.id])
+          await linkBinsOfPackages(
+            (
+              await Promise.all(
+                directPkgDirs.map(async (dir) => ({
+                  location: dir,
+                  manifest: await safeReadProjectManifestOnly(dir),
+                }))
+              )
+            )
+              .filter(({ manifest }) => manifest != null) as Array<{ location: string, manifest: DependencyManifest }>,
+            project.binsDir,
+            { warn: (message: string) => logger.info({ message, prefix: project.rootDir }) }
+          )
+        }
+      }))
+
+      if ((currentLockfile != null) && !R.equals(importerIds.sort(), Object.keys(filteredLockfile.importers).sort())) {
+        Object.assign(filteredLockfile.packages, currentLockfile.packages)
+      }
+      await writeCurrentLockfile(virtualStoreDir, filteredLockfile)
```

Comment:
The current lockfile should be written. If it will not be written, pnpm might refetch the dependencies to the virtual store

Reply:
does current lockfile refer to node_modules/.pnpm/lock.yaml?

Reply:
yes, it does